src/rust/lib/graph-editor/src/component/edge.rs

```diff
@@ -44,23 +47,280 @@ const NODE_PADDING       : f32 = node::SHADOW_SIZE;
+            let maybe_snap_position     = shape.snap_to_shape(point);
+            if let Some(snap_position) = maybe_snap_position {
+                // TODO The `SNAP_THRESHOLD` test is needed to avoid visual artifacts for
+                // circles where sometimes we might snap to a part that is invisible instead of
+                // the visible shape next to it. It might be better to find a cheaper way to avoid
+                // these.
+                let snap_distance = (snap_position - point).norm();
+                if snap_distance < SNAP_THRESHOLD && shape.contains(point){
+                    return Some(SnapData::new(snap_position,*shape))
+                }
+            }
+            None
+        });
+
+        // Return the snap that is closest to the hover.
+        snap_data
+            .map(|snap_data| ((snap_data.position - point).norm(), snap_data))
+            .min_by(|(distance_a, _), (distance_b, _)| {
+                distance_a.partial_cmp(&distance_b).unwrap_or(Ordering::Equal)
+            })
+            .map(|(_, snap_data)| snap_data)
+    }
+
+    /// Try to snap the given position to the shape and compute the perpendicular cut angle. If we
+    /// can find valid values, we return a `SplitData` struct. This can fail, if the hover is too
+    /// far from any of our sub-shapes.
+    fn get_split_data_for_hover(&self,position:Vector2<f32>,area:Area) -> Option<SplitData> {
+        let maybe_snap_data = self.snap_position_to_shape(position);
+        if let Some(snap_data) = maybe_snap_data {
+            // TODO refactor into shape
+            let base_rotation = snap_data.target_shape.display_object().rotation().z;
+            let cut_angle = snap_data.target_shape.cut_angle_for_point(snap_data.position).angle() - base_rotation;
+            Some(SplitData::new(snap_data.position,area,cut_angle))
+        } else {
+            None
+        }
+    }
+
+    /// Split the shape at the given `position` and highlight the given `area`. This might fail if
+    /// the given position is too far from the shape. In that case, splitting is disabled.
+    fn enable_hover_split(&self, position:Vector2<f32>, area:Area) {
+        if let Some(split_data) = self.get_split_data_for_hover(position,area) {
+            self.edge_shape_views().iter().for_each(|shape| shape.enable_hover_split(split_data));
+        } else {
+            self.disable_hover_split()
+        }
+    }
+
+    fn disable_hover_split(&self) {
+        for shape in self.edge_shape_views() {
+            shape.disable_hover_split()
+        }
+    }
+}
+
+
+
 // =========================
 // === Shape Definitions ===
 // =========================
 
-macro_rules! define_corner_start {($($color:tt)*) => {
+/// SplitShape allows a shape to be split along a line and each sub-shape to be colored separately.
+struct SplitShape {
+    part_a : AnyShape,
+    part_b : AnyShape,
+    joint  : AnyShape,
+}
```

Comment:
the names are hard to understand here. What is `part_a` and `part_b`? What is `joint`? Cover that in docs pls :)